Implement the Authentication module

mohitk108 opened the pull request:
Part of https://github.com/qaul/qaul.net/issues/531

Built on top of #671

mohitk108 committed:
feat(auth): add protobuf definitions for authentication messages

mohitk108 committed:
feat(auth): add basic design for auth module

mohitk108 committed:
feat(auth): implement auth challenge generation and nonce management

mohitk108 committed:
feat(auth): complete the following helper func. for auth module

- verify_challenge
- mark_authenticated
- is_authenticated
- logout
- cleanup_expired_challenge

mohitk108 committed:
fix(build): add auth.proto to protobuf generation and file copying

mohitk108 committed:
fix(mod): add auth to qaul_rpc.proto

mohitk108 committed:
feat(auth): complete the rpc handling for auth req & res

mohitk108 committed:
style: format auth.rs with rustfmt

mohitk108 committed:
chore(auth): add todo placeholders for cli implementations

also includes labelling structs and impl as dead_code to suppress warnings.

MathJud commented on rust/libqaul/src/utilities/upgrade/v2_0_0_rc_1/mod.rs:
This file must not be changed here, as it marks the upgrade procedure to qaul version 2.0.0-rc.1
We must create a new folder and a new qaul version (2.0.0 rc.4) to make the configuration file upgrade needed for your changes.
This should be done shortly before merging the PR(s) and must to reflect all changes of the configuration file.

mohitk108 replied:
resolved: dropped the commit

MathJud reviewed (changes requested):
please remove the changes in `rust/libqaul/src/utilities/upgrade/v2_0_0_rc_1/mod.rs`

MathJud reviewed (changes requested):
please remove the changes in `rust/libqaul/src/utilities/upgrade/v2_0_0_rc_1/mod.rs`

MathJud commented on rust/libqaul/src/rpc/auth.rs:
I suggest to name the File `authentication.rs` too.
It is a bit longer but in general better understandable and is the same name as the implementation.

MathJud commented on rust/libqaul/src/rpc/auth.proto:
I suggest to name the file `authentication.proto` and also use the full word Authentication in the message names.
AuthenticationRpc
AuthenticationRequest
...

mohitk108 committed:
feat(protobuf): extend the protobuf definitions to communicate salt over rpc

mohitk108 committed:
feat(libqaul): implement authentication challenge-response logic

mohitk108 committed:
feat(auth): add rpc handler for auth module messages

mohitk108 committed:
feat(users): add salt storage in configuration

mohitk108 committed:
feat(cli): add authentication module foundation

mohitk108 committed:
feat(cli): add login/logout/status commands

mohitk108 committed:
chore(libqaul): initialize auth module on starting the node

mohitk108 committed:
chore(cli): route auth module messages to Auth::rpc handler

mohitk108 committed:
chore: update the version upgrade file for libqaul

mohitk108 committed:
docs: add documentatio for authentication module in libqaul and cli client

mohitk108 committed:
docs: update cli documentation with new commands with instructions

mohitk108 committed:
refractor: rename auth files with complete authentication

MathJud commented on rust/clients/cli/README.md:
```diff
@@ -26,6 +26,13 @@ The following commands are available:
   * `account default` - get's and displays the default user account
   * `account create {User Name}` - create a new user account with the name {User Name}
+  * `account create {User Name} -p {password}` - create a new user account with a password
+  * `account create {User Name} -p` - create a new user account and prompt for password
```
account creation with only `-p` option is not working as exspected:

```
account create A2 with Password -p 
New user account created:
A2 with Password -p, ID[12D3KooWE7rnm5eZ8YWFsoXiS29BA3iFz1T6XxWrLA6WTQ1TQuPy]
    public key: 5JVd1x4qweEmTv2HoRPqPFdRVgQTWGSNNZjV5QrZWxg3
```

```
account create A2.4 -p
New user account created:
A2.4 -p, ID[12D3KooWN92C1QpyjcmdN8J4ZAHX13owBAj73HcqkLpcpYCJm8WE]
    public key: DKf2GHFGYiW8xEndvZYBEFj6gqgU1bYMnkTbSZ3QsBnJ
```

mohitk108 replied:
This is because you're not using the correct format. If you want to set your password as the string "xyz", then you have two options:
`account create alice -p xyz`
`account create alice --password xyz`

Since after the flag `-p` you don't provide any values, the CLI interprets it as the part of the `username` (note that we are allowed to use spaces in usernames)

MathJud replied:
Your documentation of the `README.md` says otherwise. Please check line 30 and adapt the documentation.

mohitk108 replied:
No, what I actually meant was you must atleast have one char after `-p` flag. But yes, that is confusing and I've improved that in commit 2c695b35a657e02b81f1f0a8fcae0a9446ef2a8a.

MathJud commented on rust/clients/cli/README.md:
```diff
@@ -26,6 +26,13 @@ The following commands are available:
   * `account create {User Name}` - create a new user account with the name {User Name}
+  * `account create {User Name} -p {password}` - create a new user account with a password
+  * `account create {User Name} -p` - create a new user account and prompt for password
+  * `account password` - set or change password for the current user account (prompts for input)
```
what happens here, when the user is not logged in?

mohitk108 replied:
His account is created, and he needs to login manually.

MathJud replied:
when invoking `account password` without the user being loged in. What is happening then?

What should happen is an information to the user: you are not logged in, please log into a user account first.

mohitk108 replied:
Thanks for pointing this out. When a user is not logged in, then user gets the information about the same.
However, earlier the user always needed to manually logout. I've changed this behavior in 1c6fac74734c22cf4a363871ce45e5814d8f1711 so that whenever we create a new account, the session would get logged out and user manually would need to login again.

MathJud commented on rust/clients/cli/README.md:
```diff
@@ -26,6 +26,13 @@ The following commands are available:
+  * `account create {User Name} -p {password}` - create a new user account with a password
+  * `account create {User Name} -p` - create a new user account and prompt for password
+  * `account password` - set or change password for the current user account (prompts for input)
+  * `account login {username} -p {password}` - login to an existing user account
```
The authentication must always take the `User ID` and not the `user name`. As only the `User ID` is unique and the sole identifyer in qaul.

mohitk108 replied:
Sure, would change it. Would like to speak more about this point with you tomorrow.

MathJud commented on rust/clients/cli/README.md:
```diff
@@ -26,6 +26,13 @@ The following commands are available:
+  * `account create {User Name} -p` - create a new user account and prompt for password
+  * `account password` - set or change password for the current user account (prompts for input)
+  * `account login {username} -p {password}` - login to an existing user account
+  * `account login {username} -p` - login and prompt for password
```
Users who have not set a password need to be able to authenticate too.
At the moment those users can not authenticate.
`account login {user ID}` without password should be an option here.

mohitk108 replied:
This makes sense. Implemented in 931a600dea8dd7a9b33017bbfbc9d8f51df55f60.

MathJud commented on rust/clients/cli/README.md:
```diff
@@ -26,6 +26,13 @@ The following commands are available:
 * user accounts
```
There needs to be an option to show all available user accounts on this machine.
The once without password must only be displayed in local settings.

mohitk108 replied:
As discussed, since this is out of scope for this PR, we'll pick this up later

MathJud commented on rust/clients/cli/README.md:
```diff
@@ -26,6 +26,13 @@ The following commands are available:
+  * `account create {User Name} -p` - create a new user account and prompt for password
+  * `account password` - set or change password for the current user account (prompts for input)
+  * `account login {username} -p {password}` - login to an existing user account
+  * `account login {username} -p` - login and prompt for password
```
When trying to login, I always get a challenge expired message:

```
account login A2.5 -p
Password: 123
Authenticating user: A2.5
Requesting users list...
Found user with password
Requesting authentication challenge...
LIBQAUL: Creating challenge for qaul_id: [0, 36, 8, 1, 18, 32, 125, 40, 230, 173, 8, 182, 49, 189, 201, 99, 109, 77, 81, 150, 108, 73, 94, 180, 21, 25, 43, 148, 185, 124, 242, 221, 33, 100, 179, 69, 91, 146]
 INFO  libqaul::rpc::authentication   > Storing challenge with key: [0, 36, 8, 1, 18, 32, 125, 40, 230, 173, 8, 182, 49, 189, 201, 99, 109, 77, 81, 150, 108, 73, 94, 180, 21, 25, 43, 148, 185, 124, 242, 221, 33, 100, 179, 69, 91, 146]
Sending authentication response...
 INFO  libqaul::rpc::authentication   > Received auth response for user_id: [0, 36, 8, 1, 18, 32, 125, 40, 230, 173, 8, 182, 49, 189, 201, 99, 109, 77, 81, 150, 108, 73, 94, 180, 21, 25, 43, 148, 185, 124, 242, 221, 33, 100, 179, 69, 91, 146]
 INFO  libqaul::rpc::authentication   > Converted to PeerId: [0, 36, 8, 1, 18, 32, 125, 40, 230, 173, 8, 182, 49, 189, 201, 99, 109, 77, 81, 150, 108, 73, 94, 180, 21, 25, 43, 148, 185, 124, 242, 221, 33, 100, 179, 69, 91, 146]
 INFO  libqaul::rpc::authentication   > Verifying challenge at timestamp: 1758540550906
 INFO  libqaul::rpc::authentication   > Looking for challenge with key: [0, 36, 8, 1, 18, 32, 125, 40, 230, 173, 8, 182, 49, 189, 201, 99, 109, 77, 81, 150, 108, 73, 94, 180, 21, 25, 43, 148, 185, 124, 242, 221, 33, 100, 179, 69, 91, 146]
 INFO  libqaul::rpc::authentication   > Total active challenges: 1
 INFO  libqaul::rpc::authentication   > Challenge key: [0, 36, 8, 1, 18, 32, 125, 40, 230, 173, 8, 182, 49, 189, 201, 99, 109, 77, 81, 150, 108, 73, 94, 180, 21, 25, 43, 148, 185, 124, 242, 221, 33, 100, 179, 69, 91, 146], expires_at: 1758540550596, current_time: 1758540550906
Authentication failed: Challenge is expired
```

mohitk108 replied:
Would look into this, since it works for me when I correctly create a user with password correctly.

```
account create rob -p xyz
New user account created:
rob, ID[12D3KooWHuT7eoFkQDnHuRY8T49HiAa2Zst7y6RqCYrTAxd4h6zo]
    public key: 965wufg3DKWoVY2hpTPwwNVC5YqUwQMMExVRnyUAoAGs
account login rob
Password required. Use: account login <username> -p <password>
account login rob -p
Password: xyz
Authenticating user: rob
Requesting users list...
Found user with password
Requesting authentication challenge...
LIBQAUL: Creating challenge for qaul_id: [0, 36, 8, 1, 18, 32, 120, 42, 254, 33, 99, 48, 236, 76, 41, 27, 225, 18, 110, 103, 86, 39, 253, 185, 225, 155, 190, 218, 225, 129, 216, 190, 41, 103, 232, 97, 180, 252]
Sending authentication response...
Authentication successful!
Session saved
```

mohitk108 replied:
Update: https://github.com/qaul/qaul.net/pull/673#discussion_r2370722154
This would make sense now.

MathJud reviewed (commented):
CLI-UI test & CLI-UI review

Please provide a step by step authentication rust unit test case, to ensure authentication is working as expected.

mohitk108 committed:
fix: flag -p not detected to prompt for password

mohitk108 committed:
fix: improvements in user experience to avoid discrepancies

- Clears a loaded session when creating a new account
- Improve message when invoked `account password` without an account logged in

mohitk108 committed:
fix: allow a user to login, if his account doesn't have password

mohitk108 committed:
feat: add non-expiring authentication system

brenodt committed:
chore: remove obsolete generated files from pre-renaming

brenodt committed:
refactor: reuse helper method hash_password
